src/features.rs

```diff
@@ -85,28 +85,72 @@ fn print_file(
     filename: Option<&str>,
 ) -> Result<()> {
     let stdin = io::stdin(); // TODO: this is not always needed
+    {
+        let reader: Box<BufRead> = match filename {
+            None => Box::new(stdin.lock()),
+            Some(filename) => Box::new(BufReader::new(File::open(filename)?)),
+        };
+
+        let highlighter = HighlightLines::new(syntax, theme);
+
+        printer.print_header(filename)?;
+
+        match printer.config.line_range.as_ref() {
+            Some(range) => {
+                print_file_ranges(printer, reader, highlighter, range)?;
+            },
+            None => {
+                print_file_no_ranges(printer, reader, highlighter)?;
+            }
+        }
+        printer.print_footer()?;
+    }
+    Ok(())
+}
 
-    let mut reader: Box<BufRead> = match filename {
-        None => Box::new(stdin.lock()),
-        Some(filename) => Box::new(BufReader::new(File::open(filename)?)),
-    };
+fn print_file_ranges<'a>(
```

Comment:
It would be really great if we had just a single function and would avoid this code duplication. I would try to change the `ranges` parameter to accept an `Option<Range>`.